Python/bltinmodule.c

```diff
@@ -46,14 +46,86 @@ _Py_IDENTIFIER(stderr);
 
 #include "clinic/bltinmodule.c.h"
 
+static PyObject*
+update_bases(PyObject* bases, PyObject** args, int nargs, int* modified_bases)
+{
+    int i, j, ind, tot_extra = 0;
+    PyObject *base, *meth, *new_base, *new_sub_base, *new_bases;
+    PyObject *stack[1] = {bases};
+    assert(PyTuple_Check(bases));
+
+    /* We have a separate cycle to calculate replacements with the idea that in
+       most cases we just scroll quickly though it and return original bases */
+    for (i = 2; i < nargs; i++){
+        base  = args[i];
+        if (PyType_Check(base)) {
+            continue;
+        }
+        if (!(meth = PyObject_GetAttrString(base, "__mro_entries__"))) {
+            if (!PyErr_ExceptionMatches(PyExc_AttributeError)) {
+                return NULL;
+            }
+            PyErr_Clear();
+            continue;
+        }
+        if (!PyCallable_Check(meth)) {
+            PyErr_SetString(PyExc_TypeError,
+                            "__mro_entries__ must be callable");
+            Py_DECREF(meth);
+            return NULL;
+        }
+        if (!(new_base = _PyObject_FastCall(meth, stack, 1))){
+            Py_DECREF(meth);
+            return NULL;
+        }
+        if (PyTuple_Check(new_base)) {
+            tot_extra += PyTuple_Size(new_base) - 1;
+        }
+        else {
+            PyErr_SetString(PyExc_TypeError,
+                            "__mro_entries__ must return a tuple");
+            Py_DECREF(meth);
+            return NULL;
+        }
+        Py_DECREF(base);
+        args[i] = new_base;
+        *modified_bases = 1;
+        Py_DECREF(meth);
+    }
+    if (!*modified_bases){
+        return bases;
+    }
+    new_bases = PyTuple_New(nargs - 2 + tot_extra);
+    ind = 0;
+    for (i = 2; i < nargs; i++) {
+        new_base = args[i];
+        if (!PyTuple_Check(new_base)) {
```

Comment:
How do you distinct the original tuple that doesn't have `__mro_entries__` from a result of `__mro_entries__`?